### PR TITLE
Support cpu quantile sketch with column-wise data split

### PR DESCRIPTION
**Pull request by rongou**
When doing quantile sketch with column-wise split data, we only need to sketch columns available locally, there is no need to synchronize the cuts across all the workers. Later on when we evaluate the splits, we also only need to evaluate on locally available columns, and then synchronize across the workers to find the global best split.

This PR also changes the `DMatrix::SliceCol` method slightly since we always split the columns into n (roughly) equal parts, so that we don't have to repeat the slice index calculation everywhere.

### CODE REVIEW

**Comment by rongou** [comment]
@trivialfis @hcho3

File: src/common/quantile.cc [hunk 1]
```diff
@@ -390,7 +396,7 @@ void SketchContainerImpl<WQSketch>::MakeCuts(HistogramCuts* cuts) {
   cuts->min_vals_.HostVector().resize(sketches_.size(), 0.0f);
   std::vector<typename WQSketch::SummaryContainer> final_summaries(reduced.size());
 
-  ParallelFor(reduced.size(), n_threads_, Sched::Guided(), [&](size_t fidx) {
```
Comment (rongou) [inline src/common/quantile.cc:393]: It's weird, the guided schedule is causing those test failures. Switching to auto fixes them. Maybe some bug in OpenMP?

Comment (rongou) [reply src/common/quantile.cc:393]: Hmm, never mind, looks like auto also fails sometimes. Probably some kind of race condition in this parallel for.

Comment (rongou) [reply src/common/quantile.cc]: Ok, finally found the root cause.

**Comment by rongou** [comment]
@trivialfis @hcho3 The CI is mostly passing now, the two failures are probably not related to this PR. Please take a look. Thanks!

File: src/common/quantile.cc [hunk 2]
```diff
@@ -283,7 +285,9 @@ void SketchContainerImpl<WQSketch>::AllReduce(
   collective::Allreduce<collective::Operation::kMax>(&n_columns, 1);
   CHECK_EQ(n_columns, sketches_.size()) << "Number of columns differs across workers";
 
-  AllreduceCategories(feature_types_, n_threads_, &categories_);
+  if (!col_split_) {
```
Comment (trivialfis) [inline src/common/quantile.cc:288]: Maybe we should `LOG(FATAL)` in all reduce categories if this is a column split matrix?

**Review by trivialfis — COMMENTED** [review]
overall looks good to me. one small question inlined in comments.

File: src/common/quantile.cc (hunk 2 above)
Comment (rongou) [reply src/common/quantile.cc:288]: `AllreduceCategories` is a freestanding function that doesn't have access to `this->col_split_`. Maybe we should make it a member function?

Comment (rongou) [reply src/common/quantile.cc]: Made it a private member function. PTAL

**Review by trivialfis — APPROVED** [review]
Looks good to me! Please merge from the latest master branch to resolve CI issue.

**Comment by rongou** [comment]
Not sure why the dask test is failing. I can't reproduce it locally. But it does look a little suspicious since it's `test_quantile_dmatrix`. Will try to dig a bit more.

**Comment by trivialfis** [comment]
It's unrelated. I have a fix.

**Pull request merged**